File: data-structures/trie.cpp
#include<bits/stdc++.h>
using namespace std;
const int tw = 26;

class trie{
  public:
    trie *next[tw];
    bool isword;

    trie(){
      for(int i=0; i<tw; i++) next[i] = NULL;
      isword = false;
    }
};

trie *head = new trie();

void insert(string word){
  trie *temp = head;
  for(int i=0; i<word.size(); i++){
    int index = word[i] - 'a';
    if(!temp->next[index]) temp->next[index] = new trie();
    temp = temp->next[index]; 
  }
  temp->isword = true;
}

bool search(string word){
  trie *temp = head;
  for(int i=0; i<word.size(); i++){
    int index = word[i] - 'a';
    if(!temp->next[index]) return false;
    temp = temp->next[index]; 
  }
  return true;
}

bool empty_next(trie* root){
  for(int i=0; i<tw; i++) if(root->next[i]) return false;
  return true;
}

trie* remove(trie *root, string word, int i){
  if(!root) return NULL;

  if(i == word.size()){
    root->isword = false;
  }
  else{
    int index = word[i] - 'a';
    root->next[index] = remove(root->next[index], word, i+1);
  }

  if(empty_next(root) && !root->isword){ delete(root); root=NULL;}

  return root;
}

int main(){
  insert("abc"); insert("nbd"); insert("nbdfg");
  head = remove(head, "abc", 0);
  cout << search("abc") << " ";
  head = remove(head, "nbd", 0);
  cout << search("nbdfg") << " ";
}

File: data-structures/trie.py
alphabets = 26

class Node:
  def __init__(self):
    self.next = [None]*alphabets
    self.is_word = False

class Trie:
  def __init__(self):
    self.root = Node()

  def insert(self, word):
    if not word: return
    temp = self.root

    for i in range(0, len(word)):
      index = ord(word[i].lower()) - ord('a')
      if not temp.next[index]: temp.next[index] = Node()
      temp = temp.next[index]

    temp.is_word = True

  def search(self, word):
    if not word: return
    temp = self.root

    for i in range(0, len(word)):
      index = ord(word[i].lower()) - ord('a')
      if not temp.next[index]: return False
      temp = temp.next[index]

    return True if temp.is_word else False

  def remove(self, word):
    if not word: return
    self._remove(self.root, word, 0)


  def _remove(self, root, word, i):
    if not root: return
    if i == len(word):
      if any(root.next):
        root.is_word = False
        return
      return True

    index = ord(word[i].lower()) - ord('a')
    delete = self._remove(root.next[index], word, i+1)

    if delete:
      del root.next[index]
      root.next[index] = None
      return None if root.is_word else True

t = Trie()
t.insert('abc')
t.insert('abcdef')
t.insert('xyz')
print('ab', t.search('ab'))
print('abcd', t.search('abcd'))
print('abc', t.search('abc'))
print('Remove abc', t.remove('abc'))
print('abc', t.search('abc'))
print('abcdef', t.search('abcdef'))
print('Remove abcdef', t.remove('abcdef'))
print('abcdef', t.search('abcdef'))
